### ISSUE TITLE
Discontinue doing synchronous validation requests upon activating plugins

### ISSUE THREAD

**Issue by westonruter** [Automattic/amp-wp#5101]
## Feature description

Depends on #4719.

Relates to #2069.

In #5100 no longer is a synchronous validation request performed when switching template modes. However, what remains is the synchronous validation request when activating plugins.

The validation errors are obtained in `AMP_Validation_Manager::validate_after_plugin_activation()` and then displayed via `AMP_Validation_Manager::print_plugin_notice()`.

Instead, upon activating a plugin when the “Plugin activated” notice is displayed:

![image](https://user-images.githubusercontent.com/134745/88474073-34a97500-ced8-11ea-90d5-55a2f3d66213.png)

There should at that point also be an admin notice from the AMP plugin which is accompanied by an enqueued script which does an _asynchronous_ request to check the site for validation issues. A spinner should be shown along with a message saying “checking for AMP compatibility issues”. Then once the results are back, a message can be shown saying there is no validation errors or that there are errors present with those activated plugins, and a link to see what they are. 

As it is currently, this admin notice should probably be suppressed if the user doesn't have developer tools enabled.

There could also be a good tie-in here with Plugin Suppression to allow a user to easily deactivate the plugins on AMP responses.

Relates to wizard compatibility scanning in #4795 and #4719.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

1. All existing server-side synchronous logic is removed.
1. Upon selecting to activate a plugin, the user is immediately taken to the resulting page with the normal "Plugin activated" notice shown.
1. On the resulting page, another "Plugin scanning" notice is displayed with a loading spinner.
1. REST API requests are initiated to do a site scan (#4719). 
1. The plugin sources for _unreviewed_ invalid markup are then aggregated from the REST API responses.
1. If there are plugin sources for unreviewed invalid markup:
   1. Give the notice a `warning` class.
   2. List out the plugins which are causing unreviewed validation errors.
   3. List out the validated URLs on which the errors occur. (The validated URLs on which a plugin's errors occur should probably be listed under the plugin itself.)
   4. Link to Plugin Suppression screen so user has ability to prevent plugins from running on AMP pages. (It's likely overkill to provide a way to suppress plugins directly from this notice.)
1. If there are no unreviewed validation errors, then the notice should be given a `success` class.

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

**Comment by jwold** [comment]
Is an initial idea with some concepts. What am I missing?

<img width="486" alt="IMG_0139" src="https://user-images.githubusercontent.com/3411173/94614589-45b98e00-025b-11eb-99c7-45a6fc3c02a3.png">

**Comment by johnwatkins0** [comment]
This looks right to me.

Side note: This will depend on the PR #5306 that I opened a few weeks ago because I think it will use the `ValidationURLProvider` created in that PR.

**Comment by westonruter** [comment]
@jwold Yes, that's right. We may want to provide information about suppressing that plugin from AMP pages as well.

I was able to reproduce the existing warning we have:

![image](https://user-images.githubusercontent.com/134745/94872834-80a9f600-0402-11eb-9e4a-50545379bfe7.png)

Here's the current code that is responsible for that:

https://github.com/ampproject/amp-wp/blob/5f42cde04d84046e106fa9784d4cf2e61f3ba204/includes/validation/class-amp-validation-manager.php#L211-L220

https://github.com/ampproject/amp-wp/blob/5f42cde04d84046e106fa9784d4cf2e61f3ba204/includes/validation/class-amp-validation-manager.php#L1972-L1993

https://github.com/ampproject/amp-wp/blob/5f42cde04d84046e106fa9784d4cf2e61f3ba204/includes/validation/class-amp-validation-manager.php#L2319-L2374

**Comment by jwold** [comment]
@westonruter are you suggesting that we allow users to disable the offending plugin directly from the notice? That could work when a single plugin is throwing errors, but might get messy if multiple plugins are mentioned in the same notice.

**Comment by westonruter** [comment]
That's true. So we should probably just add a link to go to the Plugin Suppression section so the user can decide which plugins to suppress from there.

**Comment by jwold** [comment]
From our meeting today:

This is blocked by plugin compatibility scanning. When we have the endpoint for scanning, we can delete the server side PHP code. Right now when a user activates a plugin there's a loopback request server side, this slows activation fo the plugin. Can take 15-20 seconds to activate. This has been fixed in the post editor by making validation async. 

We can do the same here, upon activating a plugin we show an admin notice that's initially PHP rendered, "checking site for compatibility," then use JS to fetch the results and put them with an admin notice loading spinner and show the results.

**Comment by kenrodriguesdesign** [comment]
Hey @westonruter,
The UX UI team discussed an alternate flow to this issue and we concluded that this flow would be better from a UX perspective.

1. Select the “Activate” action.
2. We can intercept the activation of the plugin by using this `https://developer.wordpress.org/reference/functions/register_activation_hook/`
3. Once intercepted, the AMP Plugin can check for compatibility issues.
4. Based on the results:
    1. If there are no errors then the activation of the plugin continues.
    1. If there are errors the (warning) notice will include CTA's to continue the activation or not.
    <img width="1280" alt="#5101" src="https://user-images.githubusercontent.com/61531990/129907674-7923d8e0-d91a-4dae-94ad-c51ae7d8af76.png">



If the user selects to Activate the plugin, then:
1. For technical users we can redirect them to the errors page.<img width="1280" alt="Screenshot 2021-08-18 at 7 13 38 PM" src="https://user-images.githubusercontent.com/61531990/129908986-bad53a60-cbca-4539-9b3e-e841f6f78358.png">
    


1. For non technical users, we could recommend to suppress the plugin.<img width="1281" alt="Screenshot 2021-08-18 at 7 12 05 PM" src="https://user-images.githubusercontent.com/61531990/129908974-ac442dd3-5483-4daa-affa-98a72b08d4c6.png">


I would love to know your thoughts on this, so do let me know.

**Comment by westonruter** [comment]
> 2. We can intercept the activation of the plugin by using this `https://developer.wordpress.org/reference/functions/register_activation_hook/`
> 3. Once intercepted, the AMP Plugin can check for compatibility issues.

I don't think this will be feasible because the activation hook happens when activation happens and it doesn't provide an API to intercept activation. 

If you try to output something in the activation hook, WordPress core outputs this notice:

![image](https://user-images.githubusercontent.com/134745/130155925-2c8ea5b9-7dd8-42f6-b982-767759d11e42.png)

If you try to `die` in the activation hook, then a different notice is presented:

![image](https://user-images.githubusercontent.com/134745/130155988-6e2a0c49-f0b6-4b64-87e6-35ab66898b3a.png)

In this case the plugin is not activated at all.

In order to actually evaluate whether the plugin is causing any AMP/PX issues, the plugin has to be fully active so that we can do requests on the site to see its output. 

So instead of there being a notice to continue activation after errors are encountered, it needs to instead show a notice with a "Checking for plugin compatibility…" message which is then populated asynchronously after the site scan results are obtained. The info notice would change to a success status if no errors for the activated plugins are found, or else the status would change to a warning if there were issues encountered. 

In the case of encountered errors, the user could then have the option to suppress the plugin from running on AMP pages as you've depicted, but as we discussed earlier today we want to move away from suppression as being the first choice, especially when Standard mode is selected (which we want to default to in the future). Plugin suppression is really only a viable option when in a paired AMP scheme (Transitional mode or Reader mode). If we keep the plugin suppression CTA, it may make sense just as a link to the Plugin Suppression table on the settings screen.

For technical users, the "view errors" link should probably be replaced with one or more links to Validated URL screens. If we only validate one URL after activating a plugin, then there could be just one link, but I'm thinking it will probably be best to do a full site scan (i.e. checking ~10 URLs) and then to debug the errors the user should be presented with links to each of the validated URL screens on which the activated plugin(s)' error(s) are present.

**Comment by kenrodriguesdesign** [comment]
Hey @westonruter 

Okay, I understood the flow that you've stated above and the errors that come along with it. So in this case we'll revert to the flow mentioned in the Implementation brief.

For the remaining comments:
> So instead of there being a notice to continue activation after errors are encountered, it needs to instead show a notice with a "Checking for plugin compatibility…" message which is then populated asynchronously after the site scan results are obtained. 

<img width="1280" alt="Screenshot 2021-08-25 at 5 54 45 PM" src="https://user-images.githubusercontent.com/61531990/130789981-b27cb23e-3144-449d-bfdb-fa729776a1df.png">

>The info notice would change to a success status if no errors for the activated plugins are found, or else the status would change to a warning if there were issues encountered.

<img width="1280" alt="Screenshot 2021-08-25 at 5 56 02 PM" src="https://user-images.githubusercontent.com/61531990/130790161-c79ccfbc-e832-4b6c-99eb-094b4ebd67dd.png">


> If we keep the plugin suppression CTA, it may make sense just as a link to the Plugin Suppression table on the settings screen.

Yes, at present, for non-technical users it'll be a link to the plugin suppression section in the AMP settings screen.

<img width="1280" alt="Screenshot 2021-08-25 at 5 24 00 PM" src="https://user-images.githubusercontent.com/61531990/130786032-e51ff872-11f3-4c8d-a6d6-0b222e3f4c0d.png">

For this comment:

> If we only validate one URL after activating a plugin, then there could be just one link, but I'm thinking it will probably be best to do a full site scan (i.e. checking ~10 URLs) and then to debug the errors the user should be presented with links to each of the validated URL screens on which the activated plugin(s)' error(s) are present.

I do agree with the first statement, that we could show one link for the Validated URL.
<img width="1281" alt="Screenshot 2021-08-25 at 5 50 47 PM" src="https://user-images.githubusercontent.com/61531990/130789517-8a027e52-d1db-4241-a262-a3f9bcda5d2b.png">

But for Multiple URLs i.e. site scan, showing each and every URL will not be feasible and could cause clutter.

I do have an idea for this though.
What if we show the list of URLs affected by the recently scanned plugin i.e. filter Validated URL according to the plugin. 

Do let me know if that's possible. 

If not, then would you recommend the link just directs to the list of Validated URLs?

**Comment by westonruter** [comment]
> What if we show the list of URLs affected by the recently scanned plugin i.e. filter Validated URL according to the plugin.

In the case of there being validation errors detected for the newly activated plugin(s), I think there should be (1) a link to the Plugin Suppression section, and (2) a list of links to Validated URLs on which the plugins' have errors occurring. 

The list of validated URLs could appear collapsed in a `details` element with the summary being `Found issues on X URL(s)`, and then expanding that summary could then show the individual links to the validated URLs.
 
Then after that there can be the CTA to review the plugin suppression settings.

Lastly, this would be another opportunity to mention the collection of AMP-compatible plugins. Once we have the ecosystem data integrated into the plugin directory (#4668), there could be a third CTA to "Consider alternative AMP-compatible plugins" that links to the tab in the Add Plugins screen that lists out AMP-compatible plugins.

**Comment by kenrodriguesdesign** [comment]
>In the case of there being validation errors detected for the newly activated plugin(s), I think there should be (1) a link to the Plugin Suppression section, and (2) a list of links to Validated URLs on which the plugins' have errors occurring.

Okay, understood.
But I think we should place the CTA for Plugin Suppression settings above the List of validated URLs rather than after the list.

Sending an Updated UI design for this.
For now, I have taken 3 URLs as an example, do let me know the number of URLs we decide to scan.

Collapsed view:
<img width="1281" alt="Screenshot 2021-09-01 at 4 45 29 PM" src="https://user-images.githubusercontent.com/61531990/131662173-d780888b-8739-4ea1-ba30-f97cae729493.png">

Expanded view:
<img width="1281" alt="Screenshot 2021-09-01 at 4 45 21 PM" src="https://user-images.githubusercontent.com/61531990/131662179-e4a567f8-a8bb-4bf1-b55b-a7a064f416c8.png">

Once the ecosystem data is integrated into the plugin directory, then we could have this version of the design.
 
Collapsed View:
<img width="1281" alt="Screenshot 2021-09-01 at 4 45 29 PM" src="https://user-images.githubusercontent.com/61531990/131663132-13612424-a1ba-4ada-9f4f-e7abea9319f1.png">


Expanded View:
<img width="1280" alt="Screenshot 2021-09-01 at 4 37 54 PM" src="https://user-images.githubusercontent.com/61531990/131662359-1138a336-49b9-48a4-9ea6-a5173827f45b.png">


Let me know your thoughts.

**Comment by westonruter** [comment]
Yeah, I think that's good.

> For now, I have taken 3 URLs as an example, do let me know the number of URLs we decide to scan.

The number will be variable, but figure on maybe up to 10. In any case, the list is collapsed by default so I don't think that's a problem.

**Comment by kenrodriguesdesign** [comment]
Adding a Figma link for reference and copy as well.

https://www.figma.com/file/SfMlDvHc5KHxJmAZN12PIM/AMP-Design?node-id=2903%3A22353

**Comment by fellyph** [comment]
Hi @westonruter,
 
I have tested this item the validation work as expected if I activate a valid and invalid plugin:

Valid Scenario:
![Screen Shot 2021-12-08 at 14 48 07](https://user-images.githubusercontent.com/330792/145230834-c6619538-8bd6-427f-8d89-ba109a008a8a.png)

Invalid scenario:
![Screen Shot 2021-12-08 at 14 52 38](https://user-images.githubusercontent.com/330792/145230922-2742788d-c67b-414c-8077-b403b1376832.png)

Some edge cases:

1. If I activated multiple plugins, the check doesn't mention which one was checked  
2. If I install, for example, elementor and go to review suppression, and go back to plugins list and activate a compatible plugin, I will get one confirmation and the verification from elementor appearing again, is this the behavior expected? 
  
![Screen Shot 2021-12-08 at 15 03 27](https://user-images.githubusercontent.com/330792/145231813-4863b988-a055-4034-822c-296900615e43.png)

3. When I click on "Review plugin suppression" the next page just scroll halfway to the sections:
![Screen Shot 2021-12-08 at 14 55 31](https://user-images.githubusercontent.com/330792/145232068-b3bf6042-5405-4709-bade-3559fa86c453.png)

**Comment by westonruter** [comment]
> * If I activated multiple plugins, the check doesn't mention which one was checked

@fellyph That's correct. It will return results for all plugins, not just the one(s) activated. This was something that @delawski noted as well, but there's not a clean way to determine which plugin(s) have just been activated. And it's not entirely needed anyway.

> 2\. If I install, for example, elementor and go to review suppression, and go back to plugins list and activate a compatible plugin, I will get one confirmation and the verification from elementor appearing again, is this the behavior expected?

@fellyph Are you saying that you have _suppressed_ Elementor, after activating _another_ plugin you still see validation errors coming from Elementor? I'm not seeing that. After I select “Suppressed” for Elementor in Plugin Suppression and Save:

![image](https://user-images.githubusercontent.com/134745/145345581-7dc01610-c219-4b41-bfee-047b09d59826.png)

I can activate another plugin and I don't see Elementor listed anymore:

![image](https://user-images.githubusercontent.com/134745/145345627-4fbeaaeb-5eed-49c5-91a4-ca8368c63242.png)

> 3\. When I click on "Review plugin suppression" the next page just scroll halfway to the sections:

I think this has to do with certain sections growing in height after the initial scrolling has been done. I suggest opening a low-priority issue regarding that.

**Comment by fellyph** [comment]
**QA Passed**

I will open, low priority issues for the small items
